### PR TITLE
解决@DSTransactional注解涉及到无@DS注解的mapper时会导致获取不到数据库连接

### PR DESCRIPTION
**Pull request by hzh727172424**
主要修改两个类AbstractRoutingDataSource和DynamicRoutingDataSource
AbstractRoutingDataSource的修改
    @Override
    public Connection getConnection() throws SQLException {
        String xid = TransactionContext.getXID();
        if (StringUtils.isEmpty(xid)) {
            return determineDataSource().getConnection();
        } else {
            String ds = DynamicDataSourceContextHolder.peek();
            if (StringUtils.isEmpty(ds)) {
                getConnectionProxy(determineDataSource().getConnection());
                ds = DynamicDataSourceContextHolder.peek();
            }
            ConnectionProxy connection = ConnectionFactory.getConnection(ds);
            return connection == null ? getConnectionProxy(determineDataSource().getConnection()) :
                    ConnectionFactory.getConnection(ds);
        }
    }

    @Override
    public Connection getConnection(String username, String password) throws SQLException {
        String xid = TransactionContext.getXID();
        if (StringUtils.isEmpty(xid)) {
            return determineDataSource().getConnection();
        } else {
            String ds = DynamicDataSourceContextHolder.peek();
            if (StringUtils.isEmpty(ds)) {
                getConnectionProxy(determineDataSource().getConnection(username, password));
                ds = DynamicDataSourceContextHolder.peek();
            }
            ConnectionProxy connection = ConnectionFactory.getConnection(ds);
            return connection == null ? getConnectionProxy(determineDataSource().getConnection(username, password)) :
                    ConnectionFactory.getConnection(ds);
        }
    }

    private Connection getConnectionProxy(Connection connection) {
        String ds = DynamicDataSourceContextHolder.peek();
        ConnectionProxy connectionProxy = new ConnectionProxy(connection, ds);
        ConnectionFactory.putConnection(ds, connectionProxy);
        return connectionProxy;
    }
--------------------------------------------------------------------------
DynamicRoutingDataSource类的修改
  private DataSource determinePrimaryDataSource() {
        log.debug("dynamic-datasource switch to the primary datasource");
        //解决多数据源事务当前线程获取为空的问题
        DynamicDataSourceContextHolder.push(primary);
        DataSource dataSource = dataSourceMap.get(primary);
        if (dataSource != null) {
            return dataSource;
        }
        GroupDataSource groupDataSource = groupDataSources.get(primary);
        if (groupDataSource != null) {
            return groupDataSource.determineDataSource();
        }
        throw new CannotFindDataSourceException("dynamic-datasource can not find primary datasource");
    }

### CODE REVIEW

File: src/main/java/com/baomidou/dynamic/datasource/ds/AbstractRoutingDataSource.java [hunk 1]
```diff
@@ -43,25 +43,35 @@ public Connection getConnection() throws SQLException {
...
+                    ConnectionFactory.getConnection(ds);
         }
     }
 
     @Override
     public Connection getConnection(String username, String password) throws SQLException {
         String xid = TransactionContext.getXID();
         if (StringUtils.isEmpty(xid)) {
-            return determineDataSource().getConnection(username, password);
+            return determineDataSource().getConnection();
```
Comment (huayanYu) [inline src/main/java/com/baomidou/dynamic/datasource/ds/AbstractRoutingDataSource.java:60]: 为什么要去掉username, password

File: src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java [hunk 2]
```diff
@@ -72,6 +72,8 @@ public DataSource determineDataSource() {
 
     private DataSource determinePrimaryDataSource() {
         log.debug("dynamic-datasource switch to the primary datasource");
+        //解决多数据源事务当前线程获取为空的问题
+        DynamicDataSourceContextHolder.push(primary);
         DataSource dataSource = dataSourceMap.get(primary);
```
Comment (huayanYu) [inline src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:77]: 这里不太对，DynamicDataSourceContextHolder.push(primary)不应该放在这里， 这个类是核心主类，不负责切换数据源。 只负责拿数据源。

File: src/main/java/com/baomidou/dynamic/datasource/ds/AbstractRoutingDataSource.java [hunk 3]
```diff
@@ -43,25 +43,35 @@ public Connection getConnection() throws SQLException {
...
-                    : connection;
+            return connection == null ? getConnectionProxy(determineDataSource().getConnection(username, password)) :
+                    ConnectionFactory.getConnection(ds);
         }
     }
 
-    private Connection getConnectionProxy(String ds, Connection connection) {
+    private Connection getConnectionProxy(Connection connection) {
+        String ds = DynamicDataSourceContextHolder.peek();
         ConnectionProxy connectionProxy = new ConnectionProxy(connection, ds);
```
Comment (huayanYu) [inline src/main/java/com/baomidou/dynamic/datasource/ds/AbstractRoutingDataSource.java:75]: 还是上面的问题，职责分离。

**Review by huayanYu — COMMENTED** [review]
所以解决的问题是什么？ 只加@DsTran...默认主库吗？

**Comment by hzh727172424** [comment]
> 所以解决的问题是什么？只加@dstran ...默认主库吗？

是的

File: src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java (hunk 2 above)
Comment (hzh727172424) [reply src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:77]: 你好，但是其他类没有primary该怎么做

Comment (hzh727172424) [reply src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:77]: 你好，但是其他类没有primary该怎么做

**Pull request closed without merging**